invite_user: Convert overlay to `dialog_widget`.

ganpa3 opened the pull request:
Fixes #22957.

![image](https://user-images.githubusercontent.com/58626718/200342746-d4ab1da6-1ee0-481a-a097-6a37b4cdf966.png)

alya commented:
Hm, for some reason the "Invite users" modal is not opening for me at all -- the links in the right sidebar and the gear menu change the URL to http://localhost:9991/#invite, but no modal pops up.

ganpa3 commented:
@alya Fixed the issue. Ready for review.

alya commented:
Thanks!

We should make the Cancel / Invite buttons part sticky, like it currently is, so that you can always see it even in short windows.

ganpa3 commented:
> Thanks!
> 
> We should make the Cancel / Invite buttons part sticky, like it currently is, so that you can always see it even in short windows.

#23033 is the PR for that.

alya commented:
@sahil839 Would you be up for reviewing this PR? It might be good to get this merged before @laurynmm works on disabling the invite flow in demo orgs when there is no user email.

sahil839 commented:
@ganpa3 ca you rebase the PR? It would also possibly need some CSS changes since we have removed some bootstrap CSS rules.

laurynmm commented:
I think the changes in #23529 will also need to be incorporated as well.

ganpa3 commented:
> @ganpa3 ca you rebase the PR? It would also possibly need some CSS changes since we have removed some bootstrap CSS rules.

Rebased the PR. Ready for review.

sahil839 commented:
Code looks good overall. Noticed a few issues -
- The alignment of unit and the value input for custom expiration time is not correct.
- The `Generate Invite link` is not enabled if I click on `Generate Invite link` just after opening the modal without typing anything inside email textarea.
- The gear menu is not closed after clicking on `Invite users` option.
- The `Generate Invite link` checkbox looks disabled and also the copy button for the link. This might have been the behavior before this change, so I think shouldn't be a blocker, but just thought of pointing out while we are working on this.
![Screenshot from 2023-02-02 20-29-33](https://user-images.githubusercontent.com/35494118/216375236-68e1c967-e3cc-44ed-8dba-ee928c231e5d.png)

ganpa3 commented:
@sahil839 Made the changes. Ready for review.

sahil839 reviewed (commented):
Looks good to me overall. I guess we can remove `static/templates/invite_user.hbs` file now and also there is some unused CSS for `#invite-user-link` which does not exist anymore. Would be good to check if there is some other unnecessary CSS.

ganpa3 commented:
> Looks good to me overall. I guess we can remove `static/templates/invite_user.hbs` file now and also there is some unused CSS for `#invite-user-link` which does not exist anymore. Would be good to check if there is some other unnecessary CSS.

Made the changes. Ready for review.

sahil839 commented:
The text in select elements does not look properly aligned to me, which I guess is due to change in font-size
![Screenshot from 2023-02-07 16-00-47](https://user-images.githubusercontent.com/35494118/217221738-dc8b3b16-35df-4ce5-bce6-4a6178e85852.png)
![Screenshot from 2023-02-07 16-01-50](https://user-images.githubusercontent.com/35494118/217221771-67ed189e-39e3-4e87-8139-7e2d5c0e1a58.png)

alya commented:
Thanks!

1. I think this was discussed on another PR, but the big regression from my perspective is that the buttons at the bottom are not sticky.
2. Let's make the "Generate invite link" checkbox look the same as the stream checkboxes. It's quite odd to have it look totally different.
<img width="214" alt="Screen Shot 2023-02-08 at 12 02 28 AM" src="https://user-images.githubusercontent.com/2090066/217470531-2bbc7f20-5ca4-47f0-9239-27f6d3b47098.png">

<img width="187" alt="Screen Shot 2023-02-08 at 12 09 19 AM" src="https://user-images.githubusercontent.com/2090066/217471870-43331f1e-69f6-4327-828f-e769ffeccb38.png">

The following are not regressions from `main`, but would be nice to fix while we're here:

1. I agree about the "copy" button in dark theme -- can we make it brighter?
2. Can we remove the dragging option in the bottom right of the emails text field, or else have it drag the box down? Letting it drag it sideways (the way it currently does) doesn't seem useful, as the box already takes up most of the horizontal space, and I don't see why you'd want to narrow it.
3. If it's not too hard, can we make this selection box rectangular?

<img width="142" alt="Screen Shot 2023-02-08 at 12 10 41 AM" src="https://user-images.githubusercontent.com/2090066/217472011-212f12d9-bc6d-4380-b86b-d2ab999e3465.png">

That's all I've got at the moment!

ganpa3 commented:
> I think this was discussed on another PR, but the big regression from my perspective is that the buttons at the bottom are not sticky.

This will be fixed when https://github.com/zulip/zulip/pull/23033 is merged.

> Let's make the "Generate invite link" checkbox look the same as the stream checkboxes. It's quite odd to have it look totally different.

Done.

> Can we remove the dragging option in the bottom right of the emails text field, or else have it drag the box down? Letting it drag it sideways (the way it currently does) doesn't seem useful, as the box already takes up most of the horizontal space, and I don't see why you'd want to narrow it.

Made the dragging to be vertical.

Ready for review.

alya commented:
Hm, it looks like the changes didn't get pushed?

ganpa3 commented:
> Hm, it looks like the changes didn't get pushed?

Pushed the changes. Ready for review.

alya commented:
Thanks, looks reasonable to me! @sahil839 Is this PR ready for Tim to take a look once it passes CI?

sahil839 commented:
Yes, there is nothing more from my side. Can be reviewed by Tim once CI is fixed (which shows a lint error).

timabbott commented on static/styles/zulip.css:
```diff
@@ -2336,6 +2336,7 @@ textarea.invitee_emails {
     max-height: 300px;
     width: 96%;
     max-width: 96%;
+    resize: vertical !important;
```
What's this do?

ganpa3 replied:
It makes the email text field be resized only vertically. Requested by Alya in [this comment](https://github.com/zulip/zulip/pull/23466#issuecomment-1422200962).

timabbott commented on static/js/invite.js:
```diff
@@ -270,73 +236,128 @@ function set_custom_time_inputs_visibility() {
-        }
-    });
+export function initialize() {
+    $(document).on("click", ".invite-user-link", open_invite_user_modal);
```
Sould this be a proper click handler that does `preventDefault/stopPropagation`?

ganpa3 replied:
The `open_invite_user_modal` function does make [calls](https://github.com/zulip/zulip/pull/23466/files#diff-dc7103df355c47e67a9af54d3a6de32f3d91cf8949911a0cd976fe73c907c610R240-R241) to `preventDefault/stopPropagation`

timabbott commented on static/js/invite.js:
```diff
@@ -270,73 +236,128 @@ function set_custom_time_inputs_visibility() {
+    e.stopPropagation();
+    e.preventDefault();
+
+    gear_menu.close();
```
Does this throw an exception or anything if the gear menu isn't open?

ganpa3 replied:
The `close()` is called only if the gear_menu [is open](https://github.com/zulip/zulip/blob/a2ea5e859875e033dd93c88efd0c6e0f57d07735/web/src/gear_menu.js#L152-L154)

timabbott commented:
Did a quick review; overall looks great and I think my main concern is that we definitely want the https://github.com/zulip/zulip/pull/23033 behavior to apply to this modal. @sahil839 can you work with @ganpa3 to get to a state that you feel good about merging there?

timabbott commented:
@ganpa3 it looks like your rebased this -- can you respond to the comments above?

ganpa3 commented:
@timabbott Replied to the above comments and rebased. Ready for review.

timabbott commented:
Merged, huge thanks for all the work on this @ganpa3!

@laurynmm FYI, since I seem to recall you were waiting on this to do additional changes to the modal.
